athena-gcs/src/main/java/com/amazonaws/athena/connectors/gcs/GcsSchemaUtils.java

```diff
@@ -0,0 +1,147 @@
+ * Copyright (C) 2019 - 2020 Amazon Web Services
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+package com.amazonaws.athena.connectors.gcs;
+
+import com.amazonaws.athena.connector.lambda.data.SchemaBuilder;
+import com.amazonaws.athena.connectors.gcs.storage.StorageMetadata;
+import com.amazonaws.athena.connectors.gcs.storage.datasource.StorageDatasourceConfig;
+import com.amazonaws.athena.connectors.gcs.storage.datasource.StorageTable;
+import org.apache.arrow.dataset.file.FileFormat;
+import org.apache.arrow.dataset.file.FileSystemDatasetFactory;
+import org.apache.arrow.dataset.jni.NativeMemoryPool;
+import org.apache.arrow.dataset.scanner.ScanOptions;
+import org.apache.arrow.dataset.scanner.Scanner;
+import org.apache.arrow.dataset.source.Dataset;
+import org.apache.arrow.dataset.source.DatasetFactory;
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.memory.RootAllocator;
+import org.apache.arrow.vector.ipc.ArrowReader;
+import org.apache.arrow.vector.types.TimeUnit;
+import org.apache.arrow.vector.types.Types;
+import org.apache.arrow.vector.types.pojo.ArrowType;
+import org.apache.arrow.vector.types.pojo.Field;
+import org.apache.arrow.vector.types.pojo.FieldType;
+import org.apache.arrow.vector.types.pojo.Schema;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Optional;
+
+import static com.amazonaws.athena.connectors.gcs.GcsUtil.isFieldTypeNull;
+import static com.amazonaws.athena.connectors.gcs.storage.StorageConstants.BLOCK_PARTITION_COLUMN_NAME;
+import static com.amazonaws.athena.connectors.gcs.storage.StorageUtil.createUri;
+
+public class GcsSchemaUtils
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger(GcsSchemaUtils.class);
+
+    private static final ArrowType.ArrowTypeID TIMESTAMP_NANO_TYPE = new ArrowType.Timestamp(TimeUnit.NANOSECOND, null).getTypeID();
+
+    private GcsSchemaUtils()
+    {
+    }
+
+    /**
+     * Builds the table schema based on the provided field by the retrieved instance of {@link StorageTable}
+     *
+     * @param datasource   An instance of {@link StorageMetadata}
+     * @param databaseName Name of the bucket in GCS
+     * @param tableName    Name of the storage object (file) from GCS
+     * @return An instance of {@link Schema}
+     */
+    public static Schema buildTableSchema(StorageMetadata datasource, String databaseName, String tableName) throws Exception
```

Comment:
Why do we need this? Can't we just use the Arrow Dataset API without reading the records in to read just the schema? This way you won't have to build anything, you can simply just return the Schema.

Reply:
I actually see that you're doing what I mention in AbstractStorageMetadata.java already:

line 138:
```
       BufferAllocator allocator = new RootAllocator();
        DatasetFactory factory = new FileSystemDatasetFactory(allocator,
                NativeMemoryPool.getDefault(), getFileFormat(), uri);
        // inspect schema
        return factory.inspect().getFields();
```


So is there a reason you aren't doing something similar here?

Reply:
Yes, the [GcsSchemaUtils](https://github.com/awslabs/aws-athena-query-federation/pull/843/files/f4960a5a4e2e32000ac11914879bdfefed823ee2#diff-b8cb256ec5759f01b9431fa24e85b955758c8d00ab2c4369e48c454566f4e4b4) essentially used this method to get all the fields. It just iterates through the Field collection and adds an extra partition Field.